### ISSUE TITLE
Sqlite parse error on VIRTUAL constraint

### ISSUE THREAD

**Issue by bagage** [sqlitebrowser/sqlitebrowser#917]
### Useful extra information

Similarly to #505, it looks like Sqlite has an issue when opening a database with VIRTUAL tables:

```
Sqlite parse error:  "line 1:8: unexpected token: VIRTUAL" ( "CREATE VIRTUAL TABLE TokenizerTable USING fts3tokenize(\n                    unicodesn,\n                    \"stemmer=french\"\n                )" )
Sqlite parse error:  "line 1:8: unexpected token: VIRTUAL" ( "CREATE VIRTUAL TABLE MessageSearchTable USING fts4(\n    body,\n    attachment,\n    subject,\n    from_field,\n    receivers,\n    cc,\n    bcc,\n    \n    tokenize=simple,\n    prefix=\"2,4,6,8,10\"\n)" )
```

This seems to have no effect though except of this in logs. This is a [Geary](https://github.com/GNOME/geary) mail-client database, if needed.

#### I'm opening this issue because:

- [ ] DB4S is crashing
- [x] DB4S has a bug
- [ ] DB4S needs a feature
- [ ] DB4S has another problem

#### I'm using DB4S on:

- [ ] Windows: ( _version:_ ___ )
- [x] Linux: ( _distro: Debian unstable)
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other:

#### I'm using DB4S version:

- [x] 3.9.1
- [ ] 3.9.0
- [x] Other: git master

#### I have also:

- [x] Tried out the nightly version: https://github.com/sqlitebrowser/sqlitebrowser#windows
- [x] Searched for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by justinclift** [comment]
Ahhh, that might be our grammar parser needing working then.

@MKleusberg Your kind of thing?

**Comment by MKleusberg** [comment]
Yes! I'm referencing our help-wanted issue #741 here, just because that's exactly the kind of issue I was thinking about back then :smiley:

**Comment by ghellquist** [comment]
Hi! I can see that the word VIRTUAL is not part of the list of tokens in sqlite3.g. This probably (have a bit of work to do until I understand things fully) is the reason it is not understood.

To the people that has coded this: is this a conscious design decision, or simply a missed detail? I can do a bit of testing to see if I can update the grammar with the full language description in the next few days (have a little bit of time available).

**Comment by MKleusberg** [comment]
@ghellquist Feel free to play around with this! :smiley: We highly appreciate any help in coding!

I've started playing around with this a couple of days ago, too, and it looks like this is a bit more difficult. In the grammar file you'll need to add the VIRTUAL and the USING keyword, but also add some branch for parsing the slightly different syntax of those virtual tables. In our grammar code (in src/sqlitetypes.cpp) there's more work to be done, though, to actually interpret and store the new information.

**Comment by MKleusberg** [comment]
I have added basic support for virtual tables to the grammar parser. It's still a long way to go from here because browsing and editing those tables still doesn't work as expected. But at least the ```Sqlite parse error``` is gone. You can check out the latest version in master and test yourself. And there's still a lot of work to do for better support of virtual tables, so I still want to encourage you to look into this :smiley: If you need any help just post here, write me an email, or post on Gitter.

**Comment by ghellquist** [comment]
Thank you.
I agree that the grammar thing seems the least difficult, had something in place (probably not as good as your solution).

Modules for creating virtual tables seems to each have their own grammar: the fts5 module as example has quite a few extras. So in order to be simple to use, the db Browser should know the standard modules grammars. Additionally, perhaps, the support tables created by a module could be treated somewhat differently from “normal” tables.

A bit of fun, and a long road for me get up to speed in things I haven´t done for a long time.
/Gunnar

**Comment by MKleusberg** [comment]
I've just added some code that should make it possible to browse virtual tables. Depending on the virtual table function it should also be possible to add and remove rows and edit cell values. The only thing that doesn't really work at all is editing the table schema of virtual tables, but I guess that's fine for now. Are you interested in giving tomorrow's nightly build a test, @ghellquist? I guess if it works for you we can close the issue :smiley:

**Comment by MKleusberg** [comment]
Same for you of course, @bagage! :smiley: You can check out the latest version from master and compile that again. The warning should disappear and the application should work much better with virtual tables in general.

**Comment by MKleusberg** [comment]
Has anybody had any more issues with this? If not, I'd suggest closing the issue :smile:
